### ISSUE TITLE
Search lang: filter uses file extension-only language classifier

### ISSUE THREAD

**Issue by sqs** [sourcegraph/sourcegraph#7841]
Our language classifier (`package inventory`), which is used in the GraphQL API's `language` and `languages` fields and [search stats](https://docs.sourcegraph.com/user/search#statistics), looks at both a file's name/extension *and* sometimes its content to determine the language. For example, is a `.m` file an Objective-C or Matlab file? It runs the file's content through a simple precomputed statistical model to produce the most likely language. For some languages, looking at the content is not needed; for example, `.go` files are always `Go` because there is no other language that uses that file extension.

The `lang:` and `-lang:` search filters only use the language classifier's file extension mapping, not the file content model. (Search calls [`langIncludeExcludePatterns`](https://sourcegraph.com/search?q=repo%3A^github\.com%2Fsourcegraph%2Fsourcegraph%24+langIncludeExcludePatterns) to get the list of filename patterns.) This is obviously much faster but is not correct in general (such as in the Objective-C/Matlab case above). 

One case where this is especially apparent is when using `lang:text`. The content model classifier is able to classify many more files as "Text" than can be done by file extension alone. So using `lang:text` will omit many files that are actually text files.

All of the above is a standalone inconsistency in our core search. (It's not very high priority, since it was never noticed until now and it is likely to cause false positives, which are fairly harmless in this case.)

The fix is to make `lang:` and `-lang:` use the full (file extension + content model) language classifier.

---

The problem becomes more noticeable and bad when using [search stats](https://docs.sourcegraph.com/user/search#statistics), which shows a language breakdown for all results of a query. The language breakdown uses the full (file extension + content model) language classifier. This means that if you see, for example, a simple query show `1000` lines in text files, then if you add `lang:text` it might then only show `800` lines. This occurs in [this stats query](https://sourcegraph.com/stats?q=repo:^github\.com/sourcegraph/sourcegraph$%20lang:text).

The line counts for `Text` are different:
![image](https://user-images.githubusercontent.com/1976/72583674-ffbec480-389b-11ea-960e-eb0dfccfdbcf.png)
![image](https://user-images.githubusercontent.com/1976/72583676-00eff180-389c-11ea-8d90-a4f728ab17b0.png)

**Comment by sqs** [comment]
@ryanslade I originally assigned myself because I thought this was related to the search stats page, but it's actually a shortcoming of our search. As a search problem (ignoring search stats), it's a low priority problem. It seems fairly involved to fix without hurting search perf.

@christinaforney @rvantonder Let's keep search stats experimental until it is customer-validated enough to merit the fairly involved fix to search to make this consistent.

**Comment by ryanslade** [comment]
I'm unlikely to look at this any time soon so I'm unassigning myself from it

**Comment by lguychard** [comment]
cc @sourcegraph/search-core: related to #26974

**Comment by stefanhengl** [comment]
@keegancsmith I believe you were the last one working on this. Please reopen if there is still work to do.

**Comment by keegancsmith** [comment]
We have code which solves this, but it had issues and we turned off the feature flag. We tackled this since a customer was requesting better support for apex language detection. If I remember the main issue, it was around regressions in language files which are used in multiple languages. For example C and C++ header files, if you use lang:c++ a bunch of header files that may have previously been included will now not be.

I think we just need to sit down and make a decision of what is an acceptable change in user experience here / the path to it. We also need to reunderstand the priority from the customer here, since we have let this issue sit idle for a long time. cc @jjeffwarner @eugeniaft.
